src/jwt.ts

```diff
@@ -53,37 +58,134 @@ export const decode = (token: string) => {
+        }" but was not one of "${decoded.claims.aud.join(', ')}"`
+      );
+    }
+    if (decoded.claims.aud.length > 1) {
+      if (!decoded.claims.azp) {
+        throw new Error(
+          'Authorized Party (azp) claim must be a string present in the ID token when Audience (aud) claim has multiple values'
+        );
+      }
+      if (decoded.claims.azp !== options.client_id) {
+        throw new Error(
+          `Authorized Party (azp) claim mismatch in the ID token; expected "${options.aud}", found "${decoded.claims.azp}"`
+        );
+      }
+    }
+  } else if (decoded.claims.aud !== options.aud) {
+    throw new Error(
+      `Audience (aud) claim mismatch in the ID token; expected "${options.aud}" but found "${decoded.claims.aud}"`
+    );
+  }
+
+  if (!decoded.claims.nonce) {
+    throw new Error(
+      'Nonce (nonce) claim must be a string present in the ID token'
+    );
+  }
+
   if (decoded.claims.nonce !== options.nonce) {
-    throw new Error('Invalid nonce');
+    throw new Error(
+      `Nonce (nonce) claim mismatch in the ID token; expected "${options.nonce}", found "${decoded.claims.nonce}"`
+    );
+  }
+
+  if (options.max_age && !decoded.claims.auth_time) {
+    throw new Error(
+      'Authentication Time (auth_time) claim must be a number present in the ID token when Max Age (max_age) is specified'
+    );
+  }
+
+  /* istanbul ignore next */
+  if (!decoded.claims.exp) {
+    throw new Error(
+      'Expiration Time (exp) claim must be a number present in the ID token'
+    );
+  }
+  if (!decoded.claims.iat) {
+    throw new Error(
+      'Issued At (iat) claim must be a number present in the ID token'
+    );
   }
 
   const now = new Date();
   const expDate = new Date(0);
   const iatDate = new Date(0);
   const nbfDate = new Date(0);
   const leeway = options.leeway || 60;
+  const authTimeDate = new Date(
+    parseInt(decoded.claims.auth_time) + parseInt(options.max_age) + leeway
+  );
   expDate.setUTCSeconds(decoded.claims.exp + leeway);
   iatDate.setUTCSeconds(decoded.claims.iat - leeway);
   nbfDate.setUTCSeconds(decoded.claims.nbf - leeway);
+
   if (now > expDate) {
-    throw new Error('id_token expired');
+    throw new Error(
+      `Expiration Time (exp) claim error in the ID token; current time (${now}) is after expiration time (${expDate})`
+    );
   }
   if (now < iatDate) {
-    throw new Error('id_token was issued in the future (invalid iat)');
+    throw new Error(
+      `Issued At (iat) claim error in the ID token; current time (${now}) is before issued at time (${iatDate})`
+    );
   }
   if (typeof decoded.claims.nbf !== 'undefined' && now < nbfDate) {
-    throw new Error('token is not yet valid (invalid notBefore)');
+    throw new Error(
+      `Not Before time (nbf) claim in the ID token indicates that this token can't be used just yet. Currrent time (${now}) is before ${nbfDate}`
+    );
+  }
+  if (typeof decoded.claims.auth_time !== 'undefined' && now > authTimeDate) {
```

Comment:
Wouldn't it be more readable if you wrap this inside a presence-check if, like the one on line 140 ?

Reply:
but I only want to do the `now > authTimeDate` check if auth_time is present, so it's not the same check